dali/pipeline/operator/eager_operator.h

```diff
@@ -0,0 +1,238 @@
+  ws_.Clear();
+  ws_.set_stream(cuda_stream);
+  auto output = RunImpl<GPUBackend, GPUBackend, TensorList<GPUBackend>, TensorList<GPUBackend>>(
+      inputs, kwargs);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
+  return output;
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> EagerOperator<MixedBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs,
+    CUDAStreamLease &cuda_stream) {
+  ws_.Clear();
+  ws_.set_stream(cuda_stream);
+  auto output = RunImpl<CPUBackend, GPUBackend, TensorVector<CPUBackend>, TensorList<GPUBackend>>(
+      inputs, kwargs);
+  CUDA_CALL(cudaStreamSynchronize(cuda_stream));
+  return output;
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<CPUBackend>>> EagerOperator<CPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<CPUBackend, CPUBackend>(inputs, kwargs, shared_thread_pool.get());
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> EagerOperator<GPUBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<GPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<GPUBackend, GPUBackend>(inputs, kwargs, shared_cuda_stream);
+}
+
+template <>
+template <>
+std::vector<std::shared_ptr<TensorList<GPUBackend>>> EagerOperator<MixedBackend>::Run(
+    const std::vector<std::shared_ptr<TensorList<CPUBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  return Run<CPUBackend, GPUBackend>(inputs, kwargs, shared_cuda_stream);
+}
+
+template <typename Backend>
+template <typename InBackend, typename OutBackend, typename WSInputType, typename WSOutputType>
+std::vector<std::shared_ptr<TensorList<OutBackend>>> EagerOperator<Backend>::RunImpl(
+    const std::vector<std::shared_ptr<TensorList<InBackend>>> &inputs,
+    const std::unordered_map<std::string, std::shared_ptr<TensorList<CPUBackend>>> &kwargs) {
+  // Convert and add inputs to the workspace.
+  for (size_t in_idx = 0; in_idx < inputs.size(); ++in_idx) {
+    auto tensor_in = std::make_shared<WSInputType>();
+    tensor_in->ShareData(*inputs[in_idx]);
+    SetDefaultLayoutIfNeeded(*tensor_in, op_spec_.GetSchema(), in_idx);
+    ws_.AddInput(tensor_in);
+  }
+
+  for (auto &arg : kwargs) {
+    ws_.AddArgumentInput(arg.first, arg.second);
+  }
+
+  std::vector<OutputDesc> output_desc{};
+  std::vector<std::shared_ptr<TensorList<OutBackend>>> outputs{};
+
+  outputs.reserve(num_outputs_);
+
+  for (size_t i = 0; i < num_outputs_; ++i) {
+    ws_.AddOutput(std::make_shared<WSOutputType>(batch_size_));
+  }
+
+  ws_.SetBatchSizes(batch_size_);
```

Comment:
Hmm, we probably should take the batch size from the inputs in this iteration, to support the iter-2-iter variable batch size. It is driven via the external source. Ofc user would be able to abuse it in DebugMode and run with different batch sizes among the operators in one iterations. 

The problem is that it infers the BS from external source, so we would need to run them ahead of other operators to infer the BS, store it in PipelineDebug and pass it here to be used, as it also affects generator operators that are inputless.

Reply:
For now staying with current implementation, might add support for variable batch size in the future.

Reply:
OK. I see the check is there.